## For Loop

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Python'da "for" döngüsü, bir dizi veya başka bir veri yapısı üzerinde dolaşmak veya belirli bir aralıkta işlem yapmak için kullanılan bir kontrol yapısıdır. Bu döngü, belirli bir koşul sağlandığı sürece belirli bir kod bloğunu tekrar tekrar çalıştırır.</span>

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">**"for" döngüsünün genel yapısı:**</span>

In [ ]:
for eleman in dizi:
    # Kod bloğu

<p style="font-family: 'Arial', sans-serif; font-size: 16px;">
    <strong style= "font-size:18px;">Burada:</strong>
    <br>
    <strong style = "font-family: 'Italic';">eleman</strong>, her bir döngü adımında dizi(string) içindeki elemanları temsil eden değişkenin adıdır.
    <br>
    <strong style = "font-family: 'Italic';">dizi</strong>, üzerinde dolaşılacak olan veri yapısını temsil eder. Bu veri yapısı bir liste(list), demet(tuple), dize(string) veya başka bir iterable nesne olabilir.
    <br>
    <strong style = "font-family: 'Italic';">kod bloğu</strong>, her bir döngü adımında çalıştırılacak olan kod parçalarını içerir.
</p>

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">**Bir dize üzerinde döngü:**</span>

In [5]:
dize = "Python"
for harf in dize:
    print(harf)

P
y
t
h
o
n


<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Bu kodda, "dize" adlı bir dize oluşturulur. "for" döngüsü, dizenin her bir karakterini sırayla alır ve "harf" adlı değişkene atar. Sonuç olarak, her bir karakter ekrana yazdırılır.</span>

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">**Bir liste üzerinde döngü:**</span>

In [7]:
liste = [1, 2, 3, 4, 5]
for eleman in liste:
    print(eleman)

1
2
3
4
5


<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Bu kod, "liste" adlı bir liste oluşturur. "for" döngüsü, listenin her bir elemanını sırayla alır ve "eleman" adlı değişkene atar. Ardından, her bir eleman ekrana yazdırılır.</span>

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">`range()` fonksiyonu, belirli bir aralıktaki sayıları üretmek için kullanılır. Bu fonksiyon, genellikle for döngüsü ile birlikte kullanılır. Başlangıç, bitiş ve adım parametreleri alabilir.</span>

In [9]:
# 0'dan 9'a kadar olan sayıları ekrana yazdıralım
for sayi in range(5):
    print(sayi)

0
1
2
3
4


<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Bu örnekte, `range(5)` fonksiyonu 0'dan 4'e kadar olan sayıları üretir. `for` döngüsü her bir sayıyı `sayi` adlı değişkene atar ve bu sayıları ekrana yazdırır.</span>

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">**Aynı örneği başlangıç, bitiş ve adım parametreleriyle de yapabiliriz:**</span>

In [10]:
# 1'den 10'a kadar olan çift sayıları ekrana yazdıralım
for sayi in range(2, 11, 2):
    print(sayi)

2
4
6
8
10


<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Bu örnekte, range(2, 11, 2) fonksiyonu 2'den başlayarak 10'a kadar olan sayıları üretir ve her seferinde 2'şer atlayarak ilerler. Yani, for döngüsü 2'den başlayarak 2'şer artırarak 10'a kadar olan çift sayıları ekrana yazdırır</span>

In [1]:
toplam = 0                         # "toplam" adında bir değişken oluşturulur ve değeri sıfır olarak atanır. Bu değişken, toplamın tutulacağı yerdir.

for x in range(101):               # Bu satırda bir "for" döngüsü başlatılır. "x", her döngü adımında 0'dan 100'e kadar (101 dahil değil) olan sayıları temsil eder.
    toplam += x                    # toplam += x: Döngünün her bir adımında, "toplam" değişkenine, "x" değişkeninin değeri eklenir.

print(toplam)                      # Döngü tamamlandıktan sonra, "toplam" değişkeninin son değeri ekrana yazdırılır.

5050


In [2]:
toplam = 1                  
for x in range(5):       #i değişkeni kullanılmadığı için range, döngünün 5 defa iterate(çalıştırmak) söyler
    toplam *=5           #toplam *= 5 ifadesiyle, toplam değişkeninin değeri 5 ile çarpılır
    
print(toplam)            #döngü boyunca 5'in 5. kuvveti olarak hesaplanır.

3125


<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Burada, döngü 5 kez tekrarlanır ve her adımda toplam değeri 5 ile çarpılarak güncellenir. Son olarak, toplam değişkeninin son değeri ekrana yazdırılır ve 5'in 5. kuvveti olan 3125 sonucu elde edilir.</span>

In [3]:
toplam = 1
for _ in range(5):        
    toplam *=5
    
print(toplam)       

3125


<div style="float: right; background-color: #333; padding: 5px; border: 1px solid #ccc; height: 200px; width: 800px;">
    <h3 style="color: #c0c0c0;">Not Kutusu</h3>
    <p style="font-family: 'Arial' sans-serif; font-size: 14px; color: #ffffff;">
    toplam = 1
    <br>
    for _ in range(5):        
    <br>
        toplam *=5
    <br>
    print(toplam)
    <br>
    Bu döngüde <strong>` _ `</strong> adında bir değişken kullanılmıştır. Genellikle, döngüde kullanılan bir değişkenin değerine ihtiyaç duyulmadığında  <strong>` _ `</strong> kullanılır. Yani; döngü 5 kere tekrarlanacak
    </p>
</div>

## for vs while loops

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">`for` döngüsü, eleman sayısı kadar döner ve genellikle bir dizi veya veri yapısı üzerinde gezinmek için kullanılırken, `while` döngüsü belirli bir koşul sağlandığı sürece döner ve genellikle belirli bir işlem tamamlanana kadar kullanılır.</span>

In [12]:
s = "hey"
for c in s:
    print(c)

h
e
y


<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Bu kod, "hey" dizesinin her bir karakterini alır ve ekrana basar. Döngü, dize elemanlarının sayısı kadar (3 kez) çalışır.
</span>

In [ ]:
s = "hey"
n = len(s)
index = 0

while index < n:
    print(s[index])
    index += 1

<span style="font-family: 'Arial', sans-serif; font-size: 16px;">Bu kod, bir dizenin elemanlarını for döngüsü yerine while döngüsü kullanarak alır ve ekrana basar. Burada, döngünün kaç kez döneceği, dizenin uzunluğu kadar olduğu için, döngü tam olarak 3 kez çalışır.</span>

<div style="float: right; background-color: #333; padding: 5px; border: 1px solid #ccc; height: 100x; width: 600px;">
    <h3 style="color: #c0c0c0;">Not Kutusu</h3>
    <p style="font-family: 'Arial' sans-serif; font-size: 14px; color: #ffffff;">`for` döngüsünü `while` döngüsünü kullanarak yazabiliriz, ama `while` döngüsünü `for` kullanarak yazamayız, çünkü for'da test mekanizması yok.</p>
</div>